### PR TITLE
Added policy for weighted clusters

### PR DESCRIPTION
**Pull request by jasmingacic**
Signed-off-by: jasmingacic <jasmin.gacic@gmail.com>

This PR introduces option for traffic splitting.
This PR **loosely** handles some of the requested features mentioned in https://github.com/kubeshop/kusk-gateway/issues/879 

Extended x-kusk policies with upstreams (plural) which is mutually exclusive with upstream

```
    x-kusk:
      upstreams:
      - service:
          name: httpbin1
          namespace: default
          port: 8080
          weight: 50
      - service:
          name: httpbin2
          namespace: default
          port: 8080
          weight: 50
```
This translates that traffic will be split between two upstreams by 50% https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/traffic_splitting (please help me understand if I missed anything) 

Also one unknown is `runtime_key_prefix` and `header_name`. I would appreciate any help I can get with this.
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-weightedcluster

The example above will result in following envoy config:
```
 "routes": [
         {
          "match": {
           "path": "/hello",
           "headers": [
            {
             "name": ":method",
             "string_match": {
              "safe_regex": {
               "google_re2": {},
               "regex": "^GET$|^OPTIONS$"
              }
             }
            }
           ]
          },
          "route": {
           "weighted_clusters": {
            "clusters": [
             {
              "name": "httpbin1.default.svc.cluster.local.-8080",
              "weight": 50
             },
             {
              "name": "httpbin2.default.svc.cluster.local.-8080",
              "weight": 50
             }
            ]
           },
....
 {
          "match": {
           "path": "/validated",
           "headers": [
            {
             "name": ":method",
             "string_match": {
              "safe_regex": {
               "google_re2": {},
               "regex": "^POST$|^OPTIONS$"
              }
             }
            }
           ]
          },
          "route": {
           "weighted_clusters": {
            "clusters": [
             {
              "name": "httpbin1.default.svc.cluster.local.-8080",
              "weight": 50
             },
             {
              "name": "httpbin2.default.svc.cluster.local.-8080",
              "weight": 50
             }
            ]
           },
```

### CODE REVIEW

File: smoketests/samples/weighted/weighted_api.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                    type: string
+              description: ""
+    x-kusk:
+      cors:
+        methods:
+        - GET
+        - POST
+        origins:
+        - '*'
+      upstreams:
```
Comment (mbana) [inline smoketests/samples/weighted/weighted_api.yaml:59]: Does the total `weight` of each `upstream` need to be `100`, if so, how are you going to validate this?

**Review by mbana — COMMENTED** [review]
Looking good.

I've just left some general feedback as have yet to try and run the branch.

**Comment by aabedraba** [comment]
Maybe we could add guard rails for users so the sum of the weights is always 100?

File: internal/controllers/parser.go [hunk 2]
```diff
@@ -729,8 +777,9 @@ func generateCORSPolicy(corsOpts *options.CORSOptions) (*route.CorsPolicy, error
 }
 
 type HostPortPair struct {
-	Host string
-	Port uint32
+	Host   string
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply internal/controllers/parser.go:780]: the difference in these two is the first one is returning values for internal services and the second one is returning values for hosts.
I don't see an issue here

File: pkg/options/upstream.go [hunk 3]
```diff
@@ -70,10 +73,14 @@ type UpstreamService struct {
...
 	// Port is the upstream K8s Service's port.
 	Port uint32 `yaml:"port" json:"port"`
+
+	// Weight indicates the percentage of traffic to be sent to the given upstream.
+	Weight int `yaml:"weight,omitempty" json:"weight,omitempty"`
 }
 
 func (o UpstreamHost) Validate() error {
 	return v.ValidateStruct(&o,
+		v.Field(&o.Weight),
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply pkg/options/upstream.go:83]: Done, can you look into validating combined value of weights for validation

File: internal/controllers/parser.go [hunk 4]
```diff
@@ -834,6 +871,91 @@ func generateRoute(
...
+			Name:   clusterName,
+			Weight: &wrapperspb.UInt32Value{Value: uint32(weight)},
+		})
+	} else {
+		weightedClusters = &route.WeightedCluster{
+			Clusters: []*route.WeightedCluster_ClusterWeight{
+				{
+					Name:   clusterName,
+					Weight: &wrapperspb.UInt32Value{Value: uint32(weight)},
+					ResponseHeadersToAdd: []*envoy_config_core_v3.HeaderValueOption{
```
Comment (jasmingacic) [inline internal/controllers/parser.go:942]: @mbana this is what I added so we can easily from which cluster the response is coming

File: internal/traffic/traffic.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package traffic
```
Comment (jasmingacic) [inline internal/traffic/traffic.go:23]: love it

File: pkg/options/options.go [hunk 6]
```diff
@@ -74,6 +79,13 @@ func (o Options) Validate() error {
 		return err
 	}
 
+	if o.Upstreams != nil {
```
Comment (jasmingacic) [inline pkg/options/options.go:82]: we should add a check here that will check if sum of all upstream weights is equal to 100

File: pkg/options/options.go [hunk 7]
```diff
@@ -74,6 +79,26 @@ func (o Options) Validate() error {
...
+			}
+
+			if upstream.Service != nil {
+				totalWeight = totalWeight + upstream.Service.Weight
+			} else if upstream.Host != nil {
+				totalWeight = totalWeight + upstream.Host.Weight
+			}
+		}
+		if totalWeight < 100 {
+			return fmt.Errorf("sum of upstream weights must be greater than 100. Current total weight of clusters is %d ", totalWeight)
```
Comment (jasmingacic) [inline pkg/options/options.go:98]:
```suggestion
			return fmt.Errorf("sum of upstream weights must be equal to 100. Current total weight of clusters is %d ", totalWeight)
```

File: internal/traffic/traffic.go (hunk 5 above)
Comment (mbana) [reply internal/traffic/traffic.go:23]: Sounds like a good idea.

**Comment by jasmingacic** [comment]
Had to disable check-basic for some reason it was not passing. Anyways I consider the test redundant as we are already setting up and tearing down API, Fleets and Staticroutes.

**Review by mbana — APPROVED** [review]
LGTM.

**Pull request merged**